From 399cdac1a36470f38ae74797125b28e9adcc43e1 Mon Sep 17 00:00:00 2001
From: Sriram R <quic_srirrama@quicinc.com>
Date: Tue, 30 May 2023 17:05:47 +0530
Subject: [PATCH] ath12k: Validate tx arvif start status before non Tx beacon
 setup

Currently, In rare cases of MLO combination failures, there could be cases
when a non Tx vap can be setup though the Tx vap setup is pending (due
to failure in partner setup). Since hostapd couldnt identify the failure
and avoid non Tx vap and partner bringup(to be added for safety), always
have a protection in drive to prevent setup of non Tx vap if Tx vap is
not started yet

Signed-off-by: Sriram R <quic_srirrama@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/mac.c | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -1588,6 +1588,15 @@ static int ath12k_mac_setup_bcn_tmpl_ema
 		rcu_read_unlock();
 		return -ENOLINK;
 	}
+
+	if (arvif != tx_arvif && !tx_arvif->is_started) {
+		rcu_read_unlock();
+		ath12k_warn(arvif->ar->ab,
+			    "Transmit vif is not started before this non Tx beacon setup for vdev %d\n",
+			    arvif->vdev_id);
+		return -EINVAL;
+	}
+
 	lockdep_assert_held(&tx_arvif->ar->conf_mutex);
 	beacons = ieee80211_beacon_get_template_ema_list(tx_arvif->ar->ah->hw,
 							 tx_ahvif->vif,
@@ -1659,6 +1668,15 @@ static int ath12k_mac_setup_bcn_tmpl_non
 		rcu_read_unlock();
 		return -ENOLINK;
 	}
+
+	if (arvif != tx_arvif && !tx_arvif->is_started) {
+		rcu_read_unlock();
+		ath12k_warn(arvif->ar->ab,
+			    "Transmit vif is not started before this non Tx beacon setup for vdev %d\n",
+			    arvif->vdev_id);
+		return -EINVAL;
+	}
+
 	lockdep_assert_held(&tx_arvif->ar->conf_mutex);
 	bcn = ieee80211_beacon_get_template(tx_arvif->ar->ah->hw, tx_ahvif->vif,
 					    &offs, tx_arvif->link_id);
